user_coffee_shop_query.rb

```diff
@@ -0,0 +1,22 @@
+require_relative 'distance_calculator'
+
+class UserCoffeeShopQuery
+  def initialize(coffee_shops, user)
+    @coffee_shops = coffee_shops
+    @user = user
+  end
+
+  def result
+    distances.sort_by { |coffee_shop| coffee_shop[:distance] }.first(3)
+  end
+
+  private
+
+  def distances
+    @coffee_shops.map { |coffee_shop| { name: coffee_shop.name, distance: distance(coffee_shop) } }
+  end
+
+  def distance(coffee_shop)
+    DistanceCalculator.calculate(@user, coffee_shop)
+  end
+end
```

Comment:
As a bonus challenge (only if you like challenges and you have time). Could you optimize this algorithm, IF it can be optimized better, for a huge list of coffee shops. At least more optimized, not really the best solution

Reply:
I don't know if I know what you mean. Do you think that if the `calculate` method would receive all `coffee_shops` at once, this way we wouldn't instantiate a new object for each calculation. What do you think?

Reply:
Ahh sorry, I was referring to whole class (whole algorithm). I wanted to mark the entire file

Reply:
Please take a look at my latest updates. It is pretty obvious that sorting the distances array  would be bad if the csv file contains thousands of coffee shops. But I guess I was striving for readability and elegance of the code

Reply:
:1st_place_medal: